scanner/scanner.go

```diff
@@ -0,0 +1,185 @@
+		}
+		return err
+	}
+	if sl.Skylink == "" {
+		s.staticLogger.Warnf("SweepAndLock returned a record with an empty skylink. Record hash: %s", hex.EncodeToString(sl.Hash[:]))
+		return errors.New("empty skylink")
+	}
+	inf, desc, size, scannedSize, err := s.staticClam.ScanSkylink(sl.Skylink, abort)
+	if err != nil {
+		// Scanning failed, log the error and unlock the record for another attempt.
+		s.staticLogger.Debugln(errors.AddContext(err, "scanning failed"))
+		sl.Status = database.SkylinkStatusNew
+		sl.Timestamp = time.Now().UTC()
+		err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+		if err != nil {
+			s.staticLogger.Debugln(errors.AddContext(err, "unlocking a skylink failed"))
+		}
+		return err
+	}
+	// Sanity check: scannedSize vs size.
+	if scannedSize > size {
+		s.staticLogger.Warnf("Scanned size (%d bytes) is more than the content size (%d bytes) for skylink %s", scannedSize, size, sl.Skylink)
+	}
+	// Once the scanning is complete, we no longer need the skylink. We want to
+	// remove it and only keep its hash. We don't want our database to be an
+	// index of nasty files.
+	sl.Skylink = ""
+	sl.Status = database.SkylinkStatusComplete
+	sl.Infected = inf
+	sl.InfectionDescription = desc
+	sl.Size = size
+	sl.ScannedAllContent = scannedSize == size
+	sl.ScannedAllOffsets = false
+	sl.Timestamp = time.Now().UTC()
+	err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+	if err != nil {
+		s.staticLogger.Debugln(errors.AddContext(err, "updating a skylink's status failed"))
+	}
+	return err
+}
+
+// Start launches a background task that periodically scans the database for
+// new skylink records and sends them for scanning.
+func (s Scanner) Start() {
+	// abort channel which interrupts the current scanning operation
+	abort := make(chan bool, 1)
+	// sleepLength defines how long the thread will sleep before scanning the
+	// next skylink. Its value is controlled by SweepAndScan - while we keep
+	// finding files to scan, we'll keep this sleep at zero. Once we run out of
+	// files to scan we'll reset it to its full duration of sleepBetweenScans.
+	sleepLength := sleepBetweenScans
+	go func() {
+		for {
+			numSubsequentErrs := 0
+			select {
+			case <-s.staticCtx.Done():
+				// interrupt the current scan and exit
+				abort <- true
+				return
+			case <-time.After(sleepLength):
+				err := s.SweepAndScan(abort)
+				// This was a successful call, so the number of subsequent
+				// errors is reset and we sleep for a pre-determined period in
+				// waiting for new skylinks to be uploaded.
+				if errors.Contains(err, database.ErrNoDocumentsFound) {
+					sleepLength = sleepBetweenScans
+					numSubsequentErrs = 0
+					return
+				}
+				// On error, we sleep for an increasing amount of time - from
+				// 100ms on the first error to 100s on the fourth and subsequent
+				// errors.
+				if err != nil {
+					sleepLength = sleepOnErrStep * time.Duration(math.Pow10(numSubsequentErrs))
```

Comment:
this could be extracted into a unit-tested function. e.g. `func errSleepDuration(numSubsequentErrs)`.

Reply:
I was scolded by David for not taking any shortcuts in order to expedite delivery. So, I'm taking those shortcuts now.

Should I not be doing that?

Reply:
F/U: https://github.com/SkynetLabs/malware-scanner/issues/6

Reply:
f/u